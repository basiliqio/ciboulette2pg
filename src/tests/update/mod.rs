use super::*;

mod query_params;
mod update_main_fails;
mod update_rel_fails;
mod update_rel_success;
mod update_success;

async fn test_update<'store>(
    pool: &mut sqlx::PgPool,
    query_end: &str,
    _test_name: &str,
    data: &BTreeMap<String, Vec<Uuid>>,
) -> Vec<sqlx::postgres::PgRow> {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse(format!("http://localhost{}", query_end).as_str()).unwrap();
    let first_person = data.get("peoples").unwrap().first().unwrap();
    let body = json!({
        "data": json!({
            "type": "peoples",
            "id": first_person,
            "attributes": json!({
                "first-name": "Sicnarf",
                "last-name": "Yor el",
            })
        })
    })
    .to_string();
    let body_opt = Some(body.as_str());
    const INTENTION: CibouletteIntention = CibouletteIntention::Update;

    let req_builder = CibouletteInboundRequestBuilder::new(INTENTION, &parsed_url, &body_opt);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteUpdateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_update(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let (query, args) = builder.build().unwrap();

    let raw_rows: Vec<sqlx::postgres::PgRow> = sqlx::query_with(&query, args)
        .fetch_all(&mut pool.acquire().await.unwrap())
        .await
        .unwrap();
    snapshot_table(
        &mut *pool,
        "db_snapshot_update_main_while_testing_query_params",
        &["peoples", "people-article", "favorite_color"],
    )
    .await;
    raw_rows
}

async fn test_update_many_to_one<'store>(
    pool: &mut sqlx::PgPool,
    query_end: &str,
    _test_name: &str,
    data: &BTreeMap<String, Vec<Uuid>>,
) -> Vec<sqlx::postgres::PgRow> {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse(format!("http://localhost{}", query_end).as_str()).unwrap();
    let last_person = data.get("peoples").unwrap().last().unwrap();
    let body = json!({
        "data": json!({
            "type": "peoples",
            "id": last_person,
        })
    })
    .to_string();
    let body_opt = Some(body.as_str());
    const INTENTION: CibouletteIntention = CibouletteIntention::Update;

    let req_builder = CibouletteInboundRequestBuilder::new(INTENTION, &parsed_url, &body_opt);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteUpdateRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_update(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let (query, args) = builder.build().unwrap();

    let raw_rows: Vec<sqlx::postgres::PgRow> = sqlx::query_with(&query, args)
        .fetch_all(&mut pool.acquire().await.unwrap())
        .await
        .unwrap();
    snapshot_table(
        &mut *pool,
        "db_snapshot_update_rels_while_testing_query_params",
        &["comments"],
    )
    .await;
    raw_rows
}
