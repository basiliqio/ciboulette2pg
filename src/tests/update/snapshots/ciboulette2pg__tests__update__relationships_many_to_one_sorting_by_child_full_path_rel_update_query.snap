---
source: src/tests/update/mod.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
WITH "cte_comments_update_0" AS (
  UPDATE
    "public"."comments"
  SET
    "author" = $1
  WHERE
    "public"."comments"."id" = $2 RETURNING *
),
"cte_comments_data_1" AS (
  SELECT
    "cte_comments_update_0"."id" :: TEXT AS "id",
    "cte_comments_update_0"."id" AS "main_id",
    NULL :: TEXT AS "related_id",
    NULL :: TEXT AS "related_type",
    $3 :: TEXT AS "type",
    JSON_BUILD_OBJECT($4, "cte_comments_update_0"."body") AS "data",
    "cte_comments_update_0"."author" AS "rel_author"
  FROM
    "cte_comments_update_0"
),
"cte_rel_peoples_data_2" AS (
  SELECT
    "public"."peoples"."id" :: TEXT AS "id",
    "public"."peoples"."id" AS "main_id",
    "cte_comments_data_1"."id" :: TEXT AS "related_id",
    $5 :: TEXT AS "related_type",
    $6 :: TEXT AS "type",
    NULL :: json AS "data",
    "public"."peoples"."favorite_color" AS "rel_favorite_color",
    "public"."peoples"."first-name" AS "sort_first-name"
  FROM
    "public"."peoples"
    INNER JOIN "cte_comments_data_1" ON "cte_comments_data_1"."rel_author" = "public"."peoples"."id"
),
"cte_final_main_data" AS (
  SELECT
    DISTINCT ON ("cte_rel_peoples_data_2"."main_id") "cte_rel_peoples_data_2".*,
    "cte_rel_peoples_data_2"."sort_first-name" AS "sort__first-name"
  FROM
    "cte_rel_peoples_data_2"
  ORDER BY
    "cte_rel_peoples_data_2"."main_id"
) (
  SELECT
    "cte_final_main_data"."id",
    "cte_final_main_data"."type",
    "cte_final_main_data"."data",
    "cte_final_main_data"."related_type",
    "cte_final_main_data"."related_id"
  FROM
    "cte_final_main_data"
  ORDER BY
    "cte_final_main_data"."sort__first-name" ASC
)
UNION
ALL (
  SELECT
    "cte_comments_data_1"."id",
    "cte_comments_data_1"."type",
    "cte_comments_data_1"."data",
    "cte_comments_data_1"."related_type",
    "cte_comments_data_1"."related_id"
  FROM
    "cte_comments_data_1"
);
